File: Sudoku.cpp
#include <iostream>
#include <string>
#include <vector>
#include <set>
#include <sstream>

class Grid
{
private:
    /*二维矩阵变量matrix*/
    std::vector<std::vector<int>> matrix;
public:
    /*构造函数，根据行列数生成二维矩阵*/
    Grid(int row, int col)
    {
        matrix.resize(row, std::vector<int>(col, 0));
    };

    /*构造函数，根据二维矩阵生成Grid对象*/
    Grid(const std::vector<std::vector<int>>& initMatrix)
    {
        matrix = initMatrix;
    }

    /*拷贝构造函数*/
    Grid(const Grid& initGrid)
    {
        matrix = initGrid.matrix;
    }

    /*析构函数*/
    ~Grid(){};

    /*获取某个坐标的值*/
    int getData(int row, int col){
        return matrix[row][col];
    }

    /*将matrix中(row, col)坐标的数值更改为num*/
    void CellChange(int row, int col, int num)
    {
        matrix[row][col] = num;
    }

    /*获取二维矩阵的第row行*/
    std::vector<int> getRow(int row)
    {
        return matrix[row];
    }

    /*获取二维矩阵的第col列*/
    std::vector<int> getCol(int col)
    {
        std::vector<int> data;
        for(auto &row: matrix)
        {
            data.push_back(row[col]);
        }
        return data;
    }

    /*根据坐标获取所在box的数据*/
    std::vector<int> getBox(int row, int col)
    {
        std::vector<int> data;
        int rowBegin = row / 3 * 3, colBegin = col / 3 * 3;
        for(int i=rowBegin; i<rowBegin+3; ++i)
        {
            for(int j=colBegin; j<colBegin+3; ++j)
            {
                data.push_back(matrix[i][j]);
            }
        }
        return data;
    }

    /*展示二维矩阵的数据*/
    void display()
    {
        for(auto &row: matrix)
        {
            std::cout << "| ";
            for(int &num: row){
                std::cout << num << " | ";
            }
            std::cout << std::endl;
        }
    }

    int rows() const{
        return matrix.size();
    }

    int cols() const{
        return matrix[0].size();
    }

    /*串行化*/
    std::string serialize() const {
        std::ostringstream oss;
        oss << rows() << " " << cols() << "\n"; // 首先写入行列信息
        for (const auto& row : matrix) {
            for (int val : row) {
                oss << val << " ";
            }
            oss << "\n"; // 每行结束后换行
        }
        return oss.str();
    }

    /*反串行化*/
    void deserialize(const std::string& str) {
        std::istringstream iss(str);
        int r, c;
        iss >> r >> c; // 读取行列信息
        matrix.resize(r, std::vector<int>(c, 0));

        for (int i = 0; i < r; ++i) {
            for (int j = 0; j < c; ++j) {
                iss >> matrix[i][j]; // 逐个读取矩阵元素
            }
        }
    }
};


class Sudoku
{
private:
    /*存储数据表格数据的变量cells*/
    Grid cells;
public:
    Sudoku(std::string input): cells(9, 9)
    {
        if(input.size() != 81){
            throw "The string length is invalid!";
        }
        for(int i=0;i<81;++i)
        {
            char ch = input[i];
            if(ch < '0' || ch > '9'){
                throw "The char is invalid!";
            }
            cells.CellChange(i/9, i%9, ch-'0');
        }
    }

    /*析构函数*/
    ~Sudoku(){};

    /*获取确定坐标的候选值*/
    std::vector<int> getCandidates(int row, int col)
    {
        std::vector<int> data = cells.getRow(row);
        std::set<int> rowSet(data.begin(), data.end());
        data = cells.getCol(col);
        std::set<int> colSet(data.begin(), data.end());
        data = cells.getBox(row, col);
        std::set<int> boxSet(data.begin(), data.end());

        while(!data.empty())
        {
            data.pop_back();
        }

        for(int i=1; i<=9; ++i)
        {
            if(!rowSet.count(i) && !colSet.count(i) && !boxSet.count(i)){
                data.push_back(i);
            }
        }

        return data;
    }

    void candidates_dispaly()
    {
        std::cout << "Below are the candidate values for the spaces:" << std::endl;
        for(int row=0; row<9; ++row)
        {
            for(int col=0; col<9; ++col)
            {
                if(cells.getData(row, col) == 0)
                {
                    std::vector<int> candidate = getCandidates(row, col);
                    std::cout << "(" << row+1 << ", " << col+1 << ") :";
                    for(int &num: candidate)
                    {
                        std::cout << num << " ";
                    } 
                    std::cout << std::endl;
                }
            }
        }
    }

    /*解决函数*/
    void Solution()
    {
        std::cout << "The original Sudoku table is displayed!" << std::endl;
        cells.display();
        std::cout << "The matrix is indexed from 1 to 81 from left to right and top to bottom!" << std::endl;
        candidates_dispaly();
    }
};

int main(int argc, char const *argv[])
{
    /*测试代码*/
    std::string input = "017903600000080000900000507072010430000402070064370250701000065000030000005601720";
    /*
    字符串长度不足81位报错测试：79036000000800009000005070720104300004020700643702
    字符串包含异常字符报错测试：01790360000008000090000050707201043000040207006437025070100006500003000000560172a
    其余正常测试案例：005060900000500000000807050090043078001000400840700080060204000000008000003010700
    */

    try
    {
        Sudoku answer(input);
        answer.Solution();
    }
    catch(const char *err)
    {
        std::cerr << err << '\n';
    }
    
    system("pause");
    return 0;
}

File: Sudoku.py
class Grid:
    def __init__(self, row, col):
        """构造函数，根据行列数生成二维矩阵"""
        self.matrix = [[0 for _ in range(col)] for _ in range(row)]

    def __copy__(self):
        """拷贝构造函数"""
        return Grid(self.rows(), self.cols(), self.matrix)

    def get_data(self, row, col):
        """获取某个坐标的值"""
        return self.matrix[row][col]

    def cell_change(self, row, col, num):
        """将matrix中(row, col)坐标的数值更改为num"""
        self.matrix[row][col] = num

    def get_row(self, row):
        """获取二维矩阵的第row行"""
        return self.matrix[row]

    def get_col(self, col):
        """获取二维矩阵的第col列"""
        return [row[col] for row in self.matrix]

    def get_box(self, row, col):
        """根据坐标获取所在box的数据"""
        data = []
        row_begin = (row // 3) * 3
        col_begin = (col // 3) * 3
        for i in range(row_begin, row_begin + 3):
            for j in range(col_begin, col_begin + 3):
                data.append(self.matrix[i][j])
        return data

    def display(self):
        """展示二维矩阵的数据"""
        for row in self.matrix:
            print("| " + " | ".join(map(str, row)) + " |")

    def rows(self):
        """返回矩阵行数"""
        return len(self.matrix)

    def cols(self):
        """返回矩阵列数"""
        return len(self.matrix[0]) if self.matrix else 0

    def serialize(self):
        """串行化"""
        result = f"{self.rows()} {self.cols()}\n"
        for row in self.matrix:
            result += " ".join(map(str, row)) + "\n"
        return result

    def deserialize(self, str_data):
        """反串行化"""
        lines = str_data.strip().split("\n")
        r, c = map(int, lines[0].split())
        self.matrix = [[0 for _ in range(c)] for _ in range(r)]
        for i in range(r):
            self.matrix[i] = list(map(int, lines[i + 1].split()))


class Sudoku:
    def __init__(self, input_str):
        """初始化Sudoku，输入字符串必须为81个字符"""
        self.cells = Grid(9, 9)
        if len(input_str) != 81:
            raise ValueError("The string length is invalid!")
        for i in range(81):
            ch = input_str[i]
            if ch < '0' or ch > '9':
                raise ValueError("The char is invalid!")
            self.cells.cell_change(i // 9, i % 9, int(ch))

    def get_candidates(self, row, col):
        """获取确定坐标的候选值"""
        row_data = self.cells.get_row(row)
        row_set = set(row_data)
        col_data = self.cells.get_col(col)
        col_set = set(col_data)
        box_data = self.cells.get_box(row, col)
        box_set = set(box_data)

        candidates = []
        for i in range(1, 10):
            if i not in row_set and i not in col_set and i not in box_set:
                candidates.append(i)
        return candidates

    def candidates_display(self):
        """展示每个空单元格的候选值"""
        print("Below are the candidate values for the spaces:")
        for row in range(9):
            for col in range(9):
                if self.cells.get_data(row, col) == 0:
                    candidate = self.get_candidates(row, col)
                    print(f"({row + 1}, {col + 1}) :", " ".join(map(str, candidate)))

    def solution(self):
        """展示数独解"""
        print("The original Sudoku table is displayed!")
        self.cells.display()
        print("The matrix is indexed from 1 to 81 from left to right and top to bottom!")
        self.candidates_display()


if __name__ == "__main__":
    input_str = "017903600000080000900000507072010430000402070064370250701000065000030000005601720"
    """
    字符串长度不足81位报错测试：79036000000800009000005070720104300004020700643702
    字符串包含异常字符报错测试：01790360000008000090000050707201043000040207006437025070100006500003000000560172a
    其余正常测试案例：005060900000500000000807050090043078001000400840700080060204000000008000003010700
    """
    
    try:
        sudoku = Sudoku(input_str)
        sudoku.solution()
    except ValueError as err:
        print(err)
